### ISSUE TITLE
same prediction for every user

### ISSUE THREAD

**Issue by RaphLot** [lyst/lightfm#320]
Hi,
I am really new to recommender system and to lightfm.
I am trying to see what the model I build recommender for all my users (considering all my items), but I keep getting the very same prediction (recommendation) for every single user.

Here the code I used :  
```
param = {'no_components': 314,
         'learning_schedule': 'adagrad',
         'loss': 'warp-kos',
         'learning_rate': 0.010372998003563394,
         'item_alpha': 1.0041412735137758e-06, 
         'user_alpha': 4.364266509627352e-09, 
         'max_sampled': 20,
         'num_epochs': 57}
num_epochs = param.pop("num_epochs")
model = LightFM(**param)

model = model.fit(interaction,
                  user_features=user_features,
                  item_features=item_features,
                  epochs=num_epochs,
                  num_threads=4)

mapp = dataset.mapping()
dict_user_id = mapp[0]
dict_item_id = mapp[2]

pid_array = np.arange(len(dict_item_id), dtype=np.int32)
d_user_pred = {}

for user in dict_user_id.keys():
    d_user_pred[user] = []

for index, user_id in dict_user_id.items():    
    sys.stdout.write("\rProcessing user " + str(index + 1) + "/ " + str(len(dict_user_id)))
    sys.stdout.flush()
    scores = model.predict(index, np.arange(len(dict_item_id)),
                           user_features=user_features,
                           item_features=item_features)
    top_items = item['item_id'][np.argsort(-scores)]
    d_user_pred[index] = top_items
 ```

My dataset shape is : 20.741 users and 2530 items.
Also, I have really low p@k, maybe that the problem ?

**Comment by maciejkula** [comment]
Is the model accurate? What's the MRR?

Looking at your code, what is `item`?

**Comment by RaphLot** [comment]
Here all the evalution I did:
Hybrid train AUC: 0.93
Hybrid test  AUC: 0.93
Precision@k: train 0.01374, test 0.01130. 
reciprocal_rank: train 0.05904, test 0.04969. (I suppose this is the MRR)

Item is a DataFrame where I stored all my items ids specific to my case. It's just to me a way to reconize the item.

**Comment by maciejkula** [comment]
There is nothing obviously wrong, but the model should not (and will not) normally give the same recommendations to everyone.

What are the predicted values? Does it consistently predict 0.0? Have you tried other losses and hyperparameters?

**Comment by RaphLot** [comment]
If I take the last predicted values, here what I get:
`array([ -9.96430874, -11.22552967, -12.57436848, ..., -12.96259212,
       -12.85723877, -12.80404377])`

I just tried it with the default params and the prediction change but are the still the same for everyone.


I printed the `user_embeddings`and I find them very close. Here a sample:


```
array([[ 0.19810183,  0.31938615, -0.31442386, ...,  0.3452737 ,
         0.30237758, -0.35419983],
       [ 0.2045071 ,  0.3318007 , -0.3208398 , ...,  0.3587329 ,
         0.30575866, -0.3844172 ],
       [ 0.196473  ,  0.30080256, -0.28448632, ...,  0.31968027,
         0.27496523, -0.35817078],
       ...,
       [ 0.19269793,  0.29828662, -0.2888104 , ...,  0.32078212,
         0.28065202, -0.3432573 ],
       [ 0.16556157,  0.28962615, -0.28945726, ...,  0.32600516,
         0.27702335, -0.36565372],
       [ 0.16648588,  0.28340086, -0.2805839 , ...,  0.31332952,
         0.26852673, -0.3434492 ]], dtype=float32)
```

**Comment by RaphLot** [comment]
I played with params and loss and I get now slightly diffrent recommendation for users. 

The problem, and I think it's the fact that the user_embeddings are close who causes this, is that is always the same 5 items or so. Example of what I get: 

user 0 : item 2, item 3, item 5, item 8, item44
user 1 : item 2, item 3, item 5, item 8, item44
user 2 : item 3, item 2, item 5, item 8, item44
user 3 : item 2, item 3, item 5, item 8, item44
user 4 : item 5, item 2, item 3, item 8, item44
user 5 : item 2, item 3, item 5, item 44, item 8

**Comment by maciejkula** [comment]
A couple of suggestions:

1. Train without user features.
2. Train without item features.
3. Are your users really different in the training data?
4. Maybe some items are just vastly popular.
5. Exclude items users have already seen from the returned recommendations.

**Comment by RaphLot** [comment]
thanks a lot for these suggestions.
I work on that and :
1. It's work I have different recommendations  for all users 
2.It's work I have different recommendations except for some users
3.  In global, yes. But in some cases there are a lot of features in commun  (like  the country). I saw you mentionned tf-idf in the some previous issues, but to be honest I don't know how to do it.

5. I will try this soon, thanks.

Do you know why I can't use both user and items features ?

**Comment by eveTu** [comment]
Hey, I have a similar problem. there is no other option? 
still, using user_features and get a unique prediction for every user? 
when I trained a model with the user_features the unique items for all the top 5 predictions is only 16 ( from 4,000 optional unique items) 
when I try to train without user_features the unique items for all the top 5 predictions are 1,500 which is much better also on the other evaluation metrics. 
all my features are categorical.

maybe I'm setting something in the wrong way? 
I appreciate any help 
Thanks

**Comment by WenjingZhang0214** [comment]
I want to confirm something.I trained model without user features and item features. Then for each user, I have same recommendations.Is this result right?

**Comment by clementechiu** [comment]
No, that means there's something wrong with your model or data. You should have different recommendations, otherwise using a model is pointless.
